File: care/emr/models/charge_item_definition.py
from django.db import models

from care.emr.models.base import SlugBaseModel


class ChargeItemDefinition(SlugBaseModel):
    facility = models.ForeignKey(
        "facility.Facility",
        on_delete=models.PROTECT,
    )
    version = models.IntegerField(default=1)
    status = models.CharField(max_length=255)
    title = models.CharField(max_length=255)
    slug = models.CharField(max_length=255)
    derived_from_uri = models.TextField(null=True, blank=True)
    description = models.TextField(null=True, blank=True)
    purpose = models.TextField(null=True, blank=True)
    price_components = models.JSONField(default=list)
    category = models.ForeignKey(
        "emr.ResourceCategory",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
    )

File: frontend/src/types/billing/chargeItemDefinition/chargeItemDefinition.ts
import {
  DiscountConfiguration,
  MonetaryComponent,
} from "@/types/base/monetaryComponent/monetaryComponent";
import { ResourceCategoryRead } from "@/types/base/resourceCategory/resourceCategory";
import { SlugConfig } from "@/types/base/slug/slugConfig";
import { TagConfig } from "@/types/emr/tagConfig/tagConfig";

export enum ChargeItemDefinitionStatus {
  draft = "draft",
  active = "active",
  retired = "retired",
}

export const CHARGE_ITEM_DEFINITION_STATUS_COLORS = {
  draft: "secondary",
  active: "primary",
  retired: "destructive",
} as const satisfies Record<ChargeItemDefinitionStatus, string>;

export interface ChargeItemDefinitionBase {
  id: string;
  status: ChargeItemDefinitionStatus;
  title: string;
  slug: string;
  derived_from_uri?: string;
  description?: string;
  purpose?: string;
  price_components: MonetaryComponent[];
  category: ResourceCategoryRead;
  slug_config: SlugConfig;
  tags: TagConfig[];
  can_edit_charge_item: boolean;
}

export interface ChargeItemDefinitionRead extends ChargeItemDefinitionBase {
  version?: number;
  category: ResourceCategoryRead;
}

export interface ChargeItemDefinitionCreate extends Omit<
  ChargeItemDefinitionBase,
  "id" | "category" | "slug_config" | "slug" | "tags"
> {
  slug_value: string;
  category: string;
  version?: number;
  discount_configuration: DiscountConfiguration | null;
}
